services/management/backup/backups_service.go

```diff
@@ -135,219 +69,234 @@ func (s *BackupsService) RestoreBackup(
 	ctx context.Context,
 	req *backupv1beta1.RestoreBackupRequest,
 ) (*backupv1beta1.RestoreBackupResponse, error) {
-	var params *prepareRestoreJobParams
-	var jobID, restoreID string
 
-	err := s.db.InTransactionContext(ctx, nil, func(tx *reform.TX) error {
-		var err error
-		params, err = s.prepareRestoreJob(tx.Querier, req.ServiceId, req.ArtifactId)
-		if err != nil {
-			return err
-		}
+	id, err := s.backupService.RestoreBackup(ctx, req.ServiceId, req.ArtifactId)
+	if err != nil {
+		return nil, err
+	}
 
-		restore, err := models.CreateRestoreHistoryItem(tx.Querier, models.CreateRestoreHistoryItemParams{
-			ArtifactID: req.ArtifactId,
-			ServiceID:  req.ServiceId,
-			Status:     models.InProgressRestoreStatus,
-		})
+	return &backupv1beta1.RestoreBackupResponse{
+		RestoreId: id,
+	}, nil
+}
+
+// ScheduleBackup add new backup task to scheduler.
+func (s *BackupsService) ScheduleBackup(ctx context.Context, req *backupv1beta1.ScheduleBackupRequest) (*backupv1beta1.ScheduleBackupResponse, error) {
+	var id string
+	err := s.db.InTransaction(func(tx *reform.TX) error {
+		svc, err := models.FindServiceByID(tx.Querier, req.ServiceId)
 		if err != nil {
 			return err
 		}
 
-		restoreID = restore.ID
-
-		service, err := models.FindServiceByID(tx.Querier, req.ServiceId)
+		_, err = models.FindBackupLocationByID(tx.Querier, req.LocationId)
 		if err != nil {
 			return err
 		}
 
-		var jobType models.JobType
-		var jobResultData *models.JobResultData
-		switch service.ServiceType {
+		var task scheduler.Task
+		switch svc.ServiceType {
 		case models.MySQLServiceType:
-			jobType = models.MySQLRestoreBackupJob
-			jobResultData = &models.JobResultData{
-				MySQLRestoreBackup: &models.MySQLRestoreBackupJobResult{
-					RestoreID: restoreID,
-				}}
+			task = scheduler.NewMySQLBackupTask(s.backupService, req.ServiceId, req.LocationId, req.Name, req.Description)
 		case models.MongoDBServiceType:
-			jobType = models.MongoDBRestoreBackupJob
-			jobResultData = &models.JobResultData{
-				MongoDBRestoreBackup: &models.MongoDBRestoreBackupJobResult{
-					RestoreID: restoreID,
-				}}
+			task = scheduler.NewMongoBackupTask(s.backupService, req.ServiceId, req.LocationId, req.Name, req.Description)
 		case models.PostgreSQLServiceType,
 			models.ProxySQLServiceType,
 			models.HAProxyServiceType,
 			models.ExternalServiceType:
-			return errors.Errorf("backup restore unimplemented for service type: %s", service.ServiceType)
+			return status.Errorf(codes.Unimplemented, "unimplemented service: %s", svc.ServiceType)
 		default:
-			return errors.Errorf("unsupported service type: %s", service.ServiceType)
+			return status.Errorf(codes.Unknown, "unknown service: %s", svc.ServiceType)
 		}
 
-		job, err := models.CreateJobResult(tx.Querier, params.AgentID, jobType, jobResultData)
-		if err != nil {
-			return err
+		t := req.StartTime.AsTime()
+		if t.Unix() == 0 {
+			t = time.Time{}
```

Comment:
What is the reason to replace unix zero time to empty time?

Reply:
when I received empty time from proto struct and converted to `time.Time`. `t.IsZero()` returned false.